src/main/rx/ghst_protocol.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#define GHST_RX_BAUDRATE                420000
+
+#define GHST_TX_BAUDRATE_FAST           400000
+#define GHST_TX_BAUDRATE_SLOW           115200
+#define GHST_BYTE_TIME_FAST_US          ((1000000/GHST_TX_BAUDRATE_FAST)*10)      // 10 bit words (8 data, 1 start, 1 stop)
+#define GHST_BYTE_TIME_SLOW_US          ((1000000/GHST_TX_BAUDRATE_SLOW)*10)
+#define GHST_UART_WORDLENGTH            UART_WORDLENGTH_8B
+
+typedef enum {
+    GHST_ADDR_RADIO             = 0x80,
+    GHST_ADDR_TX_MODULE_SYM     = 0x81,     // symmetrical, 400k pulses, 400k telemetry
+    GHST_ADDR_TX_MODULE_ASYM    = 0x88,     // asymmetrical, 400k pulses, 115k telemetry
+    GHST_ADDR_FC                = 0x82,
+    GHST_ADDR_GOGGLES           = 0x83,
+    GHST_ADDR_QUANTUM_TEE1      = 0x84,     // phase 2
+    GHST_ADDR_QUANTUM_TEE2      = 0x85,
+    GHST_ADDR_QUANTUM_GW1       = 0x86,
+    GHST_ADDR_5G_CLK            = 0x87,     // phase 3
+    GHST_ADDR_RX                = 0x89
+} ghstAddr_e;
+
+typedef enum {
+    // frame types 0x10 - 0x1f always include 4 primary channels, plus either 4 aux channels, 
+    // or other type-specific data. Expect types 0x14-0x1f to be added in the future, and even though
+    // not explicitly supported, the 4 primary channels should always be extracted. 
+    GHST_UL_RC_CHANS_HS4_FIRST  = 0x10,     // First frame type including 4 primary channels
+    GHST_UL_RC_CHANS_HS4_5TO8   = 0x10,     // primary 4 channel, plus CH5-8
+    GHST_UL_RC_CHANS_HS4_9TO12  = 0x11,     // primary 4 channel, plus CH9-12
+    GHST_UL_RC_CHANS_HS4_13TO16 = 0x12,     // primary 4 channel, plus CH13-16
+    GHST_UL_RC_CHANS_HS4_RSSI   = 0x13,     // primary 4 channel, plus RSSI, LQ, RF Mode, and Tx Power
+    GHST_UL_RC_CHANS_HS4_LAST   = 0x1f      // Last frame type including 4 primary channels
+} ghstUl_e;
```

Comment:
Based on the variety of the different frames, I guess we need to compute per-channel timeouts and use that for failsafe detection. I can imagine receiving `GHST_UL_RC_CHANS_HS4_FIRST` (short packet, gets through interference), but never receiving anything larger. In my book that's failsafe, with current logic we might end up not being able to disarm.

Reply:
@tonycake can you please clarify what is the rationale of not communicating the failsafe condition over GHST protocol making is potentially very unsafe?

Looks like we are expecting to have up to 16 frame types so far, 4 bits of the `type` field would be sufficient to accommodate all of those. Why not use one of the remaining 4 bits to signal a failsafe without any additional bandwidth considerations? 

GHST receiver detects the failsafe on it's own anyway (for SBUS and SRXL2 protocols), why got GHST protocol we expect the FC to figure out that signal is lost on it's own?

Reply:
Ah, a lovely lively discussion. So firstly, there are assumptions being made that the over the air protocol is the same as the Rx protocol. This would be an incorrect assumption. Ghost uses different over the air protocols based upon mode, and none of them have the issues that are being described here. 

Secondly, having receivers integrate their own logic for detecting (and communicating) failsafes is actually a little dangerous when the ultimate failsafe detection is handled by the FC firmware. Having different time delays for each to detect a failsafe is messy. 
Different types of pilots, and different applications may also define failsafe differently. For a racer, loosing 400ms worth of packets could easily be considered a failsafe, but for a long range pilot, with some autonomous control, this could easily be much longer. 
The rather simplistic design of the GHST rx protocol driver just leaves the job of failsafe detection to the FC, as I believe it should, and lets the FC (via the user) decide what a failsafe is. 

Just a bit more background, the interleaved packets on the rx protocol side, and fixed packet lengths, avoid transporting around 16 channels of unchanged servo positions, which starts to become a burden when packet rate increases, we are not running at 50Hz any more. In addition, the fix packet length works well for FCs that use the DMA engine to handle the entire packet capture process, instead of sending around individual bytes.

Reply:
We need to distinguish between **signal loss** and **failsafe procedure** - both are often referred to as "failsafe" creating confusion, while in reality they are different and independent stages of dealing with signal loss.

> So firstly, there are assumptions being made that the over the air protocol is the same as the Rx protocol.

This is not an assumption, but what's stated in the source code comment: `Packets are passed from the Ghost receiver to Betaflight as and when they arrive`, which I'm reading as "when an OTA packet arrives we'll pass it on the wire as it was received". If OTA protocol is different - it should be clearly stated to avoid any further confusion.

> Different types of pilots, and different applications may also define failsafe differently. For a racer, loosing 400ms worth of packets could easily be considered a failsafe, but for a long range pilot, with some autonomous control, this could easily be much longer.

The delay between the moment FC detects a signal loss and the moment when FC executes the failsafe procedure is different for racing and long range and it's entirely up to the FC to enforce. However, it has absolutely nothing to do with detecting signal loss for which the receiver is best equipped (having RSSI data, BER data, actual knowledge of OTA protocol fine details etc).

> Just a bit more background, the interleaved packets on the rx protocol side, and fixed packet lengths, avoid transporting around 16 channels of unchanged servo positions, which starts to become a burden when packet rate increases, we are not running at 50Hz any more.

I don't see how this is preventing the receiver to include an "I'm not receiving OTA packets" bit into the packet stream to assist FC in deciding whether it should trigger the failsafe procedure or not.

Offloading failsafe detection to the FC will also increase a delay between loosing the signal and acting on it. Imagine two scenarios (in both failsafe procedure is configured to trigger after 200ms of signal loss).

1. Receiver detects a signal loss and sends a "signal loss" flag. FC immediately realizes that the signal is lost. FC waits 200ms and executes the failsafe procedure. Total delay between signal loss and FC acting on it - 200ms.
2. Receiver detects a signal loss and stops sending the packet stream. FC waits 50ms before making a decision that it's a signal loss. FC waits another 200ms and executes the failsafe procedure. Total delay between signal loss and FC acting on it - 250ms.

50ms detection delay is assumed to tolerate a single on-wire packet being lost at 50Hz update rate and accommodate to some jitter.

Is adding a single bit to indicate a signal loss too much of an ask for GHST protocol?

Reply:
@tonycake

What assumptions can we make about on-wire protocol?

Can we expect to receive ALL of the interleaved frames (`GHST_UL_RC_CHANS_HS4_5TO8`, `GHST_UL_RC_CHANS_HS4_9TO12`, `GHST_UL_RC_CHANS_HS4_13TO16`) periodically?

What is the slowest rate for each of the interleaved frames?

For safety considerations we'll monitor each of those and not receiving at least one frame type for a certain time will be treated as a signal loss. Rationale - we want to make sure ARM/DISARM channel is always available to the FC.

Reply:
All this is just adding complexity to what should be a simple system. The very definition of 'signal loss' is not clear, even if you throw in RSSI, BER, and any other data. Signal loss means different things to different applications, as I tried to explain. 
Having multiple parts of the system try to define this, without the ability to tweak that definition, is a disaster. 
If you define 'signal loss' to be 'packet loss' then with the current protocol the FC has all that it needs. 

For the racers who have been testing Ghost since long before public release, signal loss takes on a different meaning. They are using BF blackbox logging to look for single lost packets, or small clusters that may be due to a transition through a null in an antenna radiation pattern. These small clusters of lost packets should not trigger failsafe. These guys are concerned that the feed-forward algorithms in betaflight deal with this case correctly. 

For the question about receiving all of the interleaved frames, it depends upon the mode that the system is in, and that is evolving as RF modes are added to support different applications. It is the responsibility of the firmware implementing the OTA protocol to do the right thing here. 

For your question about not receiving at least one frame for a certain time... so lets think through this, long range flight, primary concern is generally the first 4 channels, arming and disarming is something that happens at launch and landing. 
If there was a case where the FC is happily receiving the first 4 channels but skipped a few channels in the 5-8 range, where the arm switch is transferred, would this pilot expect that his quad would enter failsafe mode and potentially be lost? As a long range pilot myself this is absolutely not what I would want. 

To take the same example and apply it to a racer, we are sending over 200 packets per second, the probability that the higher numbered channels are missing for multiple packets, while the primary channels are received, is incredibly slim, so much so that it is not worth worrying about. 

But lets forget all this for a moment, and think through your request for a 'signal loss' bit. How would this work? Pilots don't want duplicated packets, so when we really detect signal loss (which we could define as a single lost packet) we don't want to send channel data along side this bit. 
Sending a packet without channel data, but with this bit set, is the same as just not sending a packet at all, the FC hasn't learned anything new. 

Defining this bit as 'a bunch of sequential  packets lost' has a similar problem, you don't want to send fabricated/duplicated channel data, and you are better off just not sending packets at all. 

So if there is really a valid requirement for a 'signal loss' bit, and someone can come up with a concrete definition which makes sense for all types of user, then I am happy to implement it. 

All this said, there is a type of failsafe which the SRXL-2 driver implements that did burn us with the Ghost implementation. That driver implements its own failsafe to deal with Rx 'reboot in flight', or just a complete receiver loss as would be the case if a wire became disconnected, or the Rx firmware hung. One of the reasons that it does this though is that the SRXL-2 protocol has two different baud rates, and if a receiver did reboot in flight, the driver needs to go and handshake again at possibly a different rate. This is not an issue for the GHST driver, where the receiver always runs at the same baud rate.

Reply:
> All this is just adding complexity to what should be a simple system.

That would be true if Ghost system would not support anything but GHST protocol, but you support S.Bus which has a failsafe condition built into the protocol. I really hope that you did not attempt to smplify things there and still set the `FAILSAFE` bit when necessary.If you do set the failsafe bit - the receiver is doing the logic already. If not - Ghost system is simply unsafe to use with the systems that depend on that bit being set.

> If there was a case where the FC is happily receiving the first 4 channels but skipped a few channels in the 5-8 range, where the arm switch is transferred, would this pilot expect that his quad would enter failsafe mode and potentially be lost? 

As a long range pilot and also systems engineer I would expect all parts of the system behaving consistently. That said, I would expect the RX to deliver me **all** channels consistently. Partial data delivery may be suitable when you expect the signal to recover soon (racing), but when you are flying at the brink of the range, partial data loss is pretty strong signal of things being bad.

> Sending a packet without channel data, but with this bit set, is the same as just not sending a packet at all, the FC hasn't learned anything new.

Totally wrong. Not sending data implies a delay on the FC side to realize that the data packets are not coming in. As you said - there is a good chance that you will fly through the outage without even noticing. Having stale data coming in with a bit to mark them as stale is another story - FC has a signal right away and can act on it if necessary. Have a look at my example with additional delay before actually doing the failsafe logic.

> So if there is really a valid requirement for a 'signal loss' bit, and someone can come up with a concrete definition which makes sense for all types of user, then I am happy to implement it.

As I said before - Ghost does signal loss detection for S.Bus and SRXL2 (at least I hope it does, because otherwise I have serious doubts about safety of that thing). Re-use the same logic and communicate the condition over GHST protocol for FC's consideration as you would do with RSSI. Racers firmware (Betaflight etc) can simply ignore the flag and do their own stuff.

If that's not good enough of a reason for you - I don't see a point in arguing this further, it's your system and your protocol and we'll have to adapt. Since you want us to do our own signal loss logic - here's #6327 as a proposed INAV's implementation of GHST failsafe. We expect to receive all flight channels consistently at least every 45ms. Anything less would be treated as signal loss. Recovery would take at least 4 good packets of each type to ensure we have consistent data for all channels.

Reply:
Sorry, this makes little sense, and just to highlight why it makes no sense, take this statement: 
    'not sending data implies a delay on the FC side to realize that the data packets are not coming in'. 

Whatever way you slice and dice this problem, you have a delay. You are waiting for something to happen (packet reception), which in marginal conditions, it doesn't happen. Handling failsafe in the receiver does nothing to reduce this delay. The receiver doesn't know that it is in a failsafe state until the failsafe has happened. 

What your suggested changes do is exactly what I have tried to avoid. You have built some arbitrary values into the GHST driver, which the user cannot modify without recompiling the code. 

The user should choose failsafe behavior based on your existing iNav CLI values, failsafe_delay for example.

As for SBus, and SRXL-2 failsafe bits, no, I do not set these, and for the same reasons I do not support the use of them. 
Both of these systems have a lot of obscure history behind them, especially with SBus, which predates modern flight controllers by many years, and communicates directly with hardware such as servos, which need to know failsafe conditions.

The division of responsibilities in a modern system should be simple, the receiver, and control link firmware should do all it can to ensure that data sent from the Tx is valid (Using FEC, CRC, etc.) and the flight controller should make decisions in a receiver-agnostic manner as to what the definition of a failsafe is. 
Anything other than this, especially hiding arbitrary timers in drivers which interface between the two is just asking for trouble. 

Now, if anyone can give me an explanation as to why these failsafe bits add any value, I am all ears, but 'preventing delays' is definitely not a reasonable explanation.

Reply:
See https://github.com/iNavFlight/inav/pull/6320#issuecomment-731599231. As the FC has no idea about the expected packet rate it can not know if it's 82ms loss of data happening already, or the next LR-mode packet should come in any millisecond now. Without the additional data that is currently not communicated in GHST protocol, FC has no choice but to assume worst case.